### DISCUSSION TITLE
Using LiteLLM in Gradio App?

### DISCUSSION THREAD

**Discussion by ZQ-Dev8** [BerriAI/litellm#631] [Q&A] [answered]
Hello! Thank you for the awesome library. I just stumbled upon LiteLLM and I want to integrate it with existing projects to remove the overhead of dealing with prompt templates. I'm attempting to integrate LiteLLM with an existing gradio app but running into issues. Is there an existing example out there for reference?

Specifically, I am seeking to modify the `inference()` function you typically find in Gradio chatbot demos to include LiteLLM prompt formatting, e.g. I want to modify the for loop below to stream the response from LiteLLM, rather than the typical Gradio way:

```
def inference(message, history):
    try:
        flattened_history = [item for sublist in history for item in sublist]
        full_message = " ".join(flattened_history + [message])
        partial_message = ""
        for token in client.text_generation(
            full_message,
            max_new_tokens=512,
            temperature=.7,
            top_k=100,
            top_p=.9,
            repetition_penalty=1.18,
            stream=True):
            partial_message += token
            yield partial_message
    except Exception as e:
        # Print the exception to the console for debugging
        print("Exception encountered:", str(e))
        # Optionally, you can yield a message to the user
        yield f"An Error occured please 'Clear' the error and try your question again"
```

**Comment by ghost** [comment]
Hey @dcruiz01 what's the issue you're facing?

**Comment by ghost** [comment, marked as answer]
```python
def inference(message, history):
    try:
        flattened_history = [item for sublist in history for item in sublist]
        full_message = " ".join(flattened_history + [message])
        messages = [{"role": "user", "content": full_message}]
        partial_message = ""
        for chunk in litellm.completion(
            model=<my-model>
            messages=messages,
            max_tokens=512,
            temperature=.7,
            top_p=.9,
            repetition_penalty=1.18,
            stream=True):
            partial_message += chunk
            yield partial_message
    except Exception as e:
        # Print the exception to the console for debugging
        print("Exception encountered:", str(e))
        # Optionally, you can yield a message to the user
        yield f"An Error occured please 'Clear' the error and try your question again"
```

**Reply by ghost** [reply]
just eyeballing, but i think the code might look something like this  @dcruiz01

**Reply by ZQ-Dev8** [reply]
Thanks for your answer @krrishdholakia! I'll give this a try and report back. I guess I didn't realize the return object of litellm.completion() could be iterated over as you showed above.

**Reply by ghost** [reply]
No worries - i've updated our getting started docs to reflect that @dcruiz01 https://docs.litellm.ai/docs/#streaming

thanks for the feedback. 

If you have a couple minutes - i'd love to learn how you're using litellm: https://calendly.com/d/4mp-gd3-k5k/litellm-1-1-onboarding-chat

**Reply by ZQ-Dev8** [reply]
Currently debugging the implementation a bit @krrishdholakia. I'm working with a remote TGI server, so I added `api_base=<ip>:<port>` to the call to `completion()`. I believe there's an issue with `partial_message += chunk`:
`Exception encountered: can only concatenate str (not "ModelResponse") to str`.
Digging into the ModelResponse object now.

I'd be happy to provide some feedback on usage - I'll sign up shortly :)

**Reply by ZQ-Dev8** [reply]
Got it: `partial_message += chunk['choices'][0]['delta']['content']` 

Works now. If you'd like, I can clean up the code and add it as a Gradio integration tutorial in LiteLLM docs, if interested @krrishdholakia.

**Reply by ghost** [reply]
@dcruiz01 that'd be great! Is your gradio app live? would love to check out the code/project

**Reply by ZQ-Dev8** [reply]
Not live, as it's nothing fancy. Just a simple frontend to test my backend with (TGI server hosted on a separate machine). I know I'll be swapping models around in the server in the future as I experiment. Having LiteLLM integrated now means I don't have to figure out the new prompt format each time, which is awesome.

Given the gradio app is bare-bones, I figure it would make a good tutorial to put in the LiteLLM docs. I can submit a PR if you like, or just make a gist or something for your review...

**Reply by ghost** [reply]
Either works - do you have a linkedin/twitter? happy to give a shoutout as well

**Reply by ZQ-Dev8** [reply]
Just submitted PR with the tutorial. Happy to change/fix as needed. Included my twitter handle in there. Thanks @krrishdholakia!

**Reply by flefevre** [reply]
Hello, I have tried your tutorial but it seems both gradio and litellm have evolved ( and that's a good thing)
Will you be able to share a new tutorial perhaps with a dedicated GitHub reposiry with a small example missing gradio and litellm?
Thanks for sharing at this Christmas period!
